node/core/approval-voting/src/lib.rs

```diff
@@ -727,20 +744,37 @@ where
 					}
 				) = approval_state;
 
-				if matches!(approval_outcome, ApprovalOutcome::Approved) {
-					let mut approvals: Vec<Action> = relay_block_hashes
-						.into_iter()
-						.map(|block_hash|
-							Action::IssueApproval(
-								candidate_hash,
-								ApprovalVoteRequest {
-									validator_index,
-									block_hash,
-								},
+				match approval_outcome {
+					ApprovalOutcome::Approved => {
+						let mut approvals: Vec<Action> = relay_block_hashes
+							.into_iter()
+							.map(|block_hash|
+								Action::IssueApproval(
+									candidate_hash,
+									ApprovalVoteRequest {
+										validator_index,
+										block_hash,
+									},
+								)
 							)
-						)
-						.collect();
-					actions.append(&mut approvals);
+							.collect();
+						actions.append(&mut approvals);
+					}
+					ApprovalOutcome::Malicious((candidate, session_index)) => {
```

Comment:
This seems like a lot of extra machinery to add instead of just sending the message from within the `if coin_flip == 0` block